# Mnist Handwritten Digits (CNN)
In this notebook, we will use CNNs.

## Loading Libraries

In [7]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Conv2D, MaxPooling2D, Dropout, Flatten  
from sklearn.model_selection import train_test_split
import numpy as np

In [8]:
(x_train, y_train) , (x_test, y_test) = mnist.load_data()
print(f'x_train shape = {x_train.shape}\nx_test shape = {x_test.shape}\ny_train shape = {y_train.shape}\ny_test shape = {y_test.shape}')

x_train shape = (60000, 28, 28)
x_test shape = (10000, 28, 28)
y_train shape = (60000,)
y_test shape = (10000,)


## Feature Scaling

In [9]:
x_train = x_train/255.0
x_test = x_test/255.0
x_train, x_test = np.expand_dims(x_train, axis=-1), np.expand_dims(x_test, axis=-1)
print(f'x_train shape = {x_train.shape}\nx_test shape = {x_test.shape}')

x_train shape = (60000, 28, 28, 1)
x_test shape = (10000, 28, 28, 1)


## Building CNN

In [11]:
model = Sequential()

# Adding convolutional layers.
model.add(Conv2D(
            filters=64, 
            kernel_size=(3,3), 
            activation='relu',
            input_shape=(28,28,1)
         ))

model.add(Conv2D(
            filters=64,
            kernel_size=(3,3),
            activation='relu'
         ))

model.add(MaxPooling2D(
            pool_size=(3,3)
         ))

model.add(Dropout(0.5))
model.add(Flatten())

# Passing the filtered convolutional layer to our ANN.
model.add(Dense(250, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 64)        640       
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 24, 24, 64)        36928     
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 8, 8, 64)          0         
_________________________________________________________________
dropout (Dropout)            (None, 8, 8, 64)          0         
_________________________________________________________________
flatten (Flatten)            (None, 4096)              0         
_________________________________________________________________
dense (Dense)                (None, 250)               1024250   
_________________________________________________________________
dense_1 (Dense)              (None, 10)                2

In [13]:
model.fit(x_train, y_train, epochs=5, batch_size=500) 

Epoch 1/5
120/120 [==============================] - 52s 431ms/step - loss: 0.0550 - accuracy: 0.9844
Epoch 2/5
120/120 [==============================] - 52s 430ms/step - loss: 0.0454 - accuracy: 0.9868
Epoch 3/5
120/120 [==============================] - 50s 421ms/step - loss: 0.0387 - accuracy: 0.9891
Epoch 4/5
120/120 [==============================] - 50s 414ms/step - loss: 0.0339 - accuracy: 0.9905
Epoch 5/5
120/120 [==============================] - 50s 415ms/step - loss: 0.0294 - accuracy: 0.9916


In [14]:
test_scores = model.evaluate(x_test, y_test)
print(test_scores)

313/313 [==============================] - 2s 8ms/step - loss: 0.0284 - accuracy: 0.9908
[0.02843763679265976, 0.9908000230789185]


In [15]:
train_scores = model.evaluate(x_train, y_train)
print(train_scores)

1875/1875 [==============================] - 15s 8ms/step - loss: 0.0168 - accuracy: 0.9960
[0.01684420555830002, 0.9960166811943054]


As we can see, testing accuracy came out to be 99.08% which is in agreement with training accuracy implying that our model does not overfit.

## Saving Model

In [16]:
model.save('digits_cnn.h5')